wave_lang/kernel/wave/utils/mapping_utils.py

```diff
@@ -260,19 +279,62 @@ def check_is_mapping_contiguous(
 
     expected_diff = [0] * len(index_mapping)
     expected_diff[-1] = 1
+    if expected_diff == diff:
+        return True
 
-    return diff == expected_diff
+    # If the expected pattern is not found, check if the mapping still produces
+    # contiguous memory accesses by computing offsets for each element.
+    #
+    # The check works by:
+    # 1. Computing the linear memory offset symbolically for the first element
+    # 2. For each subsequent element:
+    #    - Updating the fastest dimension in the index for that element
+    #    - Transforming through the mapping
+    #    - Computing the new memory offset
+    #    - Verifying the offset increased by exactly 1
+    # 3. Returns True only if all elements are sequential
+    fastest_dim = list(index.keys())[get_fastest_index(index)]
+
+    idxc = IndexingContext.current()
+    strides = strides_from_symbolic_shape(idxc, array_shape, allow_mixed_shapes=True)
+    new_index = transform_index_on_mapping(
+        mapping, symbolic_shape, index, is_read=is_read
+    )
+    prev_offset = _compute_offset(
+        [new_index[infer_dim(d)] for d in symbolic_shape], strides
+    )
+    for i in range(1, elements_per_thread):
+        new_index = deepcopy(index)
+        new_index[fastest_dim].start += i
+        new_index = transform_index_on_mapping(
```

Comment:
Instead of transforming the index multiple times, can we not do this just once with an index like: OLD_INDEX + V -> TRANSFORMED_INDEX and then substitute 1, 2, ..., elements_per_thread into this TRANSFORMED_INDEX for V?

Reply:
I don't think it will work, e.g. for the igemm case we are transforming index `m, n, k` -> `N, C, H, W` so no original symbols remains and we no longer know what symbol corresponds to the vector fastest dim. We need to substitute pre-transform.